editor/editor_node.cpp

```diff
@@ -891,6 +891,7 @@ void EditorNode::_resources_changed(const Vector<String> &p_resources) {
 
 	if (changed.size()) {
 		for (Ref<Resource> &res : changed) {
+			ResourceCache::erase_ref(res->get_path());
```

Comment:
Is this really necessary? It suggests that `reload_from_file()` does not work reliably.

Reply:
Yes, it is; Otherwise this error will occur:
https://github.com/godotengine/godot/blob/93cdacbb0a30f12b2f3f5e8e06b90149deeb554b/core/io/resource.cpp#L67-L77

Reply:
But `reload_from_file()` does not call `set_path()` anywhere 🤔

Reply:
>	godot.windows.editor.dev.x86_64.mono.exe!Resource::set_path(const String & p_path, bool p_take_over) Line 69	C++
 	godot.windows.editor.dev.x86_64.mono.exe!ResourceFormatLoaderCSharpScript::load(const String & p_path, const String & p_original_path, Error * r_error, bool p_use_sub_threads, float * r_progress, ResourceFormatLoader::CacheMode p_cache_mode) Line 2861	C++
 	godot.windows.editor.dev.x86_64.mono.exe!ResourceLoader::_load(const String & p_path, const String & p_original_path, const String & p_type_hint, ResourceFormatLoader::CacheMode p_cache_mode, Error * r_error, bool p_use_sub_threads, float * r_progress) Line 261	C++
 	godot.windows.editor.dev.x86_64.mono.exe!ResourceLoader::_thread_load_function(void * p_userdata) Line 319	C++
 	godot.windows.editor.dev.x86_64.mono.exe!ResourceLoader::_load_start(const String & p_path, const String & p_type_hint, ResourceLoader::LoadThreadMode p_thread_mode, ResourceFormatLoader::CacheMode p_cache_mode) Line 499	C++
 	godot.windows.editor.dev.x86_64.mono.exe!ResourceLoader::load(const String & p_path, const String & p_type_hint, ResourceFormatLoader::CacheMode p_cache_mode, Error * r_error) Line 416	C++
 	godot.windows.editor.dev.x86_64.mono.exe!Resource::reload_from_file() Line 194	C++
 	godot.windows.editor.dev.x86_64.mono.exe!EditorNode::_resources_changed(const Vector<String> & p_resources) Line 900	C++
 	godot.windows.editor.dev.x86_64.mono.exe!call_with_variant_args_helper<EditorNode,Vector<String> const &,0>(EditorNode * p_instance, void(EditorNode::*)(const Vector<String> &) p_method, const Variant * * p_args, Callable::CallError & r_error, IndexSequence<0> __formal) Line 303	C++
 	godot.windows.editor.dev.x86_64.mono.exe!call_with_variant_args<EditorNode,Vector<String> const &>(EditorNode * p_instance, void(EditorNode::*)(const Vector<String> &) p_method, const Variant * * p_args, int p_argcount, Callable::CallError & r_error) Line 418	C++
 	godot.windows.editor.dev.x86_64.mono.exe!CallableCustomMethodPointer<EditorNode,Vector<String> const &>::call(const Variant * * p_arguments, int p_argcount, Variant & r_return_value, Callable::CallError & r_call_error) Line 105	C++
 	godot.windows.editor.dev.x86_64.mono.exe!Callable::callp(const Variant * * p_arguments, int p_argcount, Variant & r_return_value, Callable::CallError & r_call_error) Line 58	C++
 	godot.windows.editor.dev.x86_64.mono.exe!Object::emit_signalp(const StringName & p_name, const Variant * * p_args, int p_argcount) Line 1127	C++
 	godot.windows.editor.dev.x86_64.mono.exe!Node::emit_signalp(const StringName & p_name, const Variant * * p_args, int p_argcount) Line 3608	C++
 	godot.windows.editor.dev.x86_64.mono.exe!Object::emit_signal<Vector<String>>(const StringName & p_name, Vector<String> <p_args_0>) Line 920	C++
 	godot.windows.editor.dev.x86_64.mono.exe!EditorFileSystem::_update_scan_actions() Line 705	C++
 	godot.windows.editor.dev.x86_64.mono.exe!EditorFileSystem::_notification(int p_what) Line 1262	C++
 	godot.windows.editor.dev.x86_64.mono.exe!EditorFileSystem::_notificationv(int p_notification, bool p_reversed) Line 146	C++
 	godot.windows.editor.dev.x86_64.mono.exe!Object::notification(int p_notification, bool p_reversed) Line 839	C++
 	godot.windows.editor.dev.x86_64.mono.exe!SceneTree::_process_group(SceneTree::ProcessGroup * p_group, bool p_physics) Line 954	C++
 	godot.windows.editor.dev.x86_64.mono.exe!SceneTree::_process(bool p_physics) Line 1031	C++
 	godot.windows.editor.dev.x86_64.mono.exe!SceneTree::process(double p_time) Line 510	C++
 	godot.windows.editor.dev.x86_64.mono.exe!Main::iteration() Line 3613	C++
 	godot.windows.editor.dev.x86_64.mono.exe!OS_Windows::run() Line 1474	C++
 	godot.windows.editor.dev.x86_64.mono.exe!widechar_main(int argc, wchar_t * * argv) Line 182	C++
 	godot.windows.editor.dev.x86_64.mono.exe!_main() Line 204	C++
 	godot.windows.editor.dev.x86_64.mono.exe!main(int argc, char * * argv) Line 218	C++
 	godot.windows.editor.dev.x86_64.mono.exe!WinMain(HINSTANCE__ * hInstance, HINSTANCE__ * hPrevInstance, char * lpCmdLine, int nCmdShow) Line 232	C++
 	[External Code]

Reply:
That looks like a bug in C# loader. Changing a core class is not the correct fix.
https://github.com/godotengine/godot/blob/4363ae34fa3f52223ae78ff2f2c187ab99827e3a/modules/mono/csharp_script.cpp#L2848-L2859
This code creates a new instance of the resource and forces it onto the path. This is wrong, as it doesn't replace the existing instances. They will point to the same path, but contain old data.

CC @godotengine/dotnet

Reply:
It seems when this code was added GDScript was doing the same thing in their loader:

https://github.com/godotengine/godot/blob/e36fb95c50ce0cd0ab9621afe668332895712c2e/modules/gdscript/gd_script.cpp#L1835-L1843

It was changed in https://github.com/godotengine/godot/pull/40598.

> This is wrong, as it doesn't replace the existing instances. They will point to the same path, but contain old data.

I'm confused about this, there should only be one `CSharpScript` instance for each C# class, so if a script instance already existed it should have been retrieved instead of creating a new one.

Reply:
Yep, it should. And this code creates a new instance every time a script is reloaded.

Reply:
I don't see how the code creates a new instance every time, the `GetOrCreateScriptBridgeForPath` method should retrieve an existing script instance for the given path if one already exists.

Reply:
Ok I'm not familiar with this code. So this
```C++
if (GDMonoCache::godot_api_cache_updated) {
		GDMonoCache::managed_callbacks.ScriptManagerBridge_GetOrCreateScriptBridgeForPath(&p_path, &scr);
```
reuses the script from resource cache? If so, the `set_path()` is just in a wrong place. It should be like this:
```C++	
if (GDMonoCache::godot_api_cache_updated) {
	GDMonoCache::managed_callbacks.ScriptManagerBridge_GetOrCreateScriptBridgeForPath(&p_path, &scr);
} else {
	scr = Ref<CSharpScript>(memnew(CSharpScript));
	scr->set_path(p_original_path);
}
```

Reply:
When a new C# script is created it's stored in a dictionary on C#'s side so it can be retrieved later by its path and avoid recreating it.

Moving the `set_path()` call makes sense to me, but that `else` shouldn't be reached anyway. GDScript seems to be doing the same thing, when a script with the same path doesn't exist in its cache it creates what they call a _shallow_ script and they set the path.

https://github.com/godotengine/godot/blob/9144457484f9d2f53990a0aac37caff1c9012e6d/modules/gdscript/gdscript_cache.cpp#L279-L294

https://github.com/godotengine/godot/blob/9144457484f9d2f53990a0aac37caff1c9012e6d/modules/gdscript/gdscript_cache.cpp#L254-L257

A few differences from what we are doing:
- When setting the script's path, they set the `p_take_over` parameter to true.
- They use `p_path` instead of `p_original_path`.